pkg/registry/core/service/ipallocator/cidrallocator.go

```diff
@@ -0,0 +1,454 @@
+	// Per example, if we have allocator A and B with 256 and 1024 possible
+	// addresses each, the chances to get B has to be 4 times the chances to
+	// get A so we can spread the load of IPs randomly.
+	// However, we need to validate the best strategy before going to Beta.
+	isIPv6 := c.ipFamily == api.IPFamily(v1.IPv6Protocol)
+	for _, allocator := range c.tree.TopLevelPrefixes(isIPv6) {
+		ip, err := allocator.AllocateNext()
+		if err == nil {
+			return ip, nil
+		}
+	}
+	return nil, ErrFull
+}
+
+func (c *MetaAllocator) Release(ip net.IP) error {
+	allocator, err := c.getAllocator(ip)
+	if err != nil {
+		return err
+	}
+	return allocator.Release(ip)
+
+}
+func (c *MetaAllocator) ForEach(f func(ip net.IP)) {
+	ipLabelSelector := labels.Set(map[string]string{
+		networkingv1alpha1.LabelIPAddressFamily: string(c.IPFamily()),
+		networkingv1alpha1.LabelManagedBy:       ControllerName,
+	}).AsSelectorPreValidated()
+	ips, err := c.ipAddressLister.List(ipLabelSelector)
+	if err != nil {
+		return
+	}
+	for _, ip := range ips {
+		f(netutils.ParseIPSloppy(ip.Name))
+	}
+}
+
+func (c *MetaAllocator) CIDR() net.IPNet {
+	return net.IPNet{}
+
+}
+func (c *MetaAllocator) IPFamily() api.IPFamily {
+	return c.ipFamily
+}
+func (c *MetaAllocator) Has(ip net.IP) bool {
+	allocator, err := c.getAllocator(ip)
+	if err != nil {
+		return false
+	}
+	return allocator.Has(ip)
+}
+func (c *MetaAllocator) Destroy() {
+	select {
+	case <-c.internalStopCh:
+	default:
+		close(c.internalStopCh)
+	}
+}
+
+// for testing
+func (c *MetaAllocator) Used() int {
+	ipLabelSelector := labels.Set(map[string]string{
+		networkingv1alpha1.LabelIPAddressFamily: string(c.IPFamily()),
+		networkingv1alpha1.LabelManagedBy:       ControllerName,
+	}).AsSelectorPreValidated()
+	ips, err := c.ipAddressLister.List(ipLabelSelector)
+	if err != nil {
+		return 0
+	}
+	return len(ips)
+}
+
+// for testing
+func (c *MetaAllocator) Free() int {
+	c.muTree.Lock()
+	defer c.muTree.Unlock()
+
+	size := 0
+	isIPv6 := c.ipFamily == api.IPFamily(v1.IPv6Protocol)
+	for _, allocator := range c.tree.TopLevelPrefixes(isIPv6) {
+		size += int(allocator.size)
+	}
+	return size - c.Used()
+}
+
+func (c *MetaAllocator) EnableMetrics() {}
+
+// DryRun returns a random allocator
+func (c *MetaAllocator) DryRun() Interface {
+	c.muTree.Lock()
+	defer c.muTree.Unlock()
+	isIPv6 := c.ipFamily == api.IPFamily(v1.IPv6Protocol)
+	for _, allocator := range c.tree.TopLevelPrefixes(isIPv6) {
+		return allocator.DryRun()
+	}
+	return &Allocator{}
+}
+
+func isReady(serviceCIDR *networkingv1alpha1.ServiceCIDR) bool {
+	if serviceCIDR == nil {
+		return false
+	}
+
+	for _, condition := range serviceCIDR.Status.Conditions {
+		if condition.Type == networkingv1alpha1.ServiceCIDRConditionReady {
+			return condition.Status == metav1.ConditionStatus(metav1.ConditionTrue)
+		}
+	}
+	// if condition is not found assume it is ready to handle scenarios
+	// where the kube-controller-manager is not running
```

Comment:
> Or is it not possible for the "add Service" and the "delete ServiceCIDR" to be overlapping from the perspective of the apiserver?

actually I described this case in the KEP and didn't implement it https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1880-multiple-service-cidrs#service-ip-reservation, instead of removing the finalizer immediately it delays the removal 

let me fix it tomorrow

Reply:
I will add all of this in a follow up